CenterForOpenScience/osf.io
# Private/View Only Rework [#OSF-5792]

Note: see https://github.com/CenterForOpenScience/osf.io/pull/4509 for original. This PR is for the rework per some requested changes. All changes in the linked PR apply to this one as well.

## Purpose
Make view-only links actually view-only, make anonymized view-only links not show public components. 

## Changes
Massively strip functionality from the project page if an active view-only link is on the URL:
1. Rework the nav bar to explain view-only link and give a way to return to the dashboard
2. Remove access to all editing functionality
3. Remove all menu bar info except Files and Wiki
4. Anonymize public components if on an anonymous link. 
5. Only allow access to components selected when creating the link, even if the component is public. This makes it so nothing is implicitly added to the view-only link unless the creator of the link makes it explicit that they wanted to include it. 
6. Logged in users and anonymous users now see exactly the same view of a project when viewing it through a view only link. 
7. If a view only link is active but a person is on something not within the scope of that link they will receive a redirect and the current view_only link will be removed. This can happen 1) if not in a project and you try a view only link, 2) if on a node that the view_only link does not contain, 3) use an invalid view_only link

## Side Effects
1. ~~One strange situation could still occur if someone were to add the view-only link to a different OSF page. Currently it will ONLY strip functionality from projects, but it will replace the top menu bar with the view only bar. I'm not sure there is a good way to change this.~~  This has been modified so that view_only links now redirect to not include the view_only link if the link is not valid

## QA Considerations
1. Both Public and Private components should be tested for view access.
2. Both Logged in (with access) and un-logged in users should be tested
3. Both anonymized and non-anonymized links should be tested
4. There should be no way to modify the project while the link is active
5. Users should not be able to gain access to other project information using a private link on a different project (Whenever I test locally it redirects to login, but I can't confirm how it will work on staging). 
6. Private links with institutions should not show institution access (I wasn't sure how to test this locally, but the code looked right)
7. Testing that view_only links redirect properly when the link is not valid for the current url location should be tested.
